ppexpect.py

```diff
@@ -0,0 +1,90 @@
+# SPDX-License-Identifier: GPL-2.0-only
+
+# Poor Python Expect.
+# Simplified expect to run a process
+
+import errno
+import fcntl
+import os
+import select
+import subprocess
+import sys
+import time
+
+__all__ = ['Process', 'TimeoutError']
+
+def set_nonblocking(fd):
+    fl = fcntl.fcntl(fd, fcntl.F_GETFL)
+    fcntl.fcntl(fd, fcntl.F_SETFL, fl | os.O_NONBLOCK)
+
+if sys.version_info[0] < 3:
+    class TimeoutError(OSError):
+        def __init__(self):
+            super(TimeoutError, self).__init__(errno.ETIMEDOUT, None)
+else:
+    class TimeoutError(TimeoutError):
+        pass
+
+class Process:
+    def __init__(self, cmd):
+        process = subprocess.Popen(
+            cmd,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            close_fds=True,
+            universal_newlines=True,
+            )
+        set_nonblocking(process.stdout.fileno())
+
+        self.process = process
+
+    def __del__(self):
+        self.close()
+
+    def write(self, msg):
+        stdin = self.process.stdin
+        stdin.write(msg)
+        stdin.flush()
+
+    def expect(self, msg, timeout=1.0):
+        data = b''
+        start = time.time()
```

Comment:
I think it would be simpler if you used a deadline instead.

So deadline = time.time() + timeout

Then pass deadline into __readTimeout().

In __readTimeout(), you can check `if now > deadline`, and pass in `deadline - now` into select().

Reply:
Yes, similar comment, usually you want to handle timeout <= 0 as no timeout, but for the "poor" implementation not supporting it and using a deadline instead is easier.